Implement network management modules

lemoony opened the pull request:
Partly addresses #102 . 

This PR basically allows to create a network and attaching apps to it.

Added two new library files: 
- library/dokku_network.py 
- library/dokku_network_property.py 

Updated README.md

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# from ansible.module_utils.basic import *
```
remove?

ltalirz commented on README.md:
```diff
@@ -502,6 +502,69 @@ Enable or disable the letsencrypt plugin for a dokku app
+
+- name: Setting a global network property
+  dokku_network_property:
+    app: --global
```
This seems a bit hackish, maybe have a look at `dokku_domains` and how this switch is handled there?

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+DOCUMENTATION = """
+---
+module: dokku_network_property
+short_description: Set or clear a network property for a given dokku application
```
Could you include an example of clearing a network property?

lemoony replied:
There actually is already an example - please see _"De-associate the container with the network."_

However, I understand that switching between the expressions _"assoication a network"_ and _"setting a network proprety"_ is a bit confusing. However, I tried to stick to the official dokku documentation when it comes to that.

What do you think about changing **"De-associate the container with the network."** to _"Clearing a network property (de-associate the container with the network)."_?

lemoony replied:
There is  already an example (see _De-associate the container with the network._) but I will rename it to make it more clear

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+    aliases: []
+  network:
+    description:
+      - The name of the network to attach to
```
```suggestion
      - The value of the network property (leave empty to unset)
```

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+
+    command = "dokku network:set {0} {1} {2}".format(
+        data["app"],
+        data["property"] if "property" in data else "",
```
Is this the right logic?

I'm reading [here](https://dokku.com/docs/networking/network/#attaching-an-app-to-a-network) that one should leave the value empty in order to de-associate the container from the network rather than leave the name of the property empty.

lemoony replied:
Yes, you're completely right. I mixed that up

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+    aliases: []
+  property:
+    description:
+      - The attach network property
```
This is just a suggestion, perhaps not the best
```suggestion
      - The network property of the app to be modified
```

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+    required: True
+    default: null
+    aliases: []
+  network:
```
Looking at [how `network:set` works](https://dokku.com/docs/networking/network/#attaching-an-app-to-a-network), I think it would be cleaner to call this variable `value`.

The network name just so happens to be the value for the three properties you have chosen to support here

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+        "network": {"required": True, "type": "str"},
+        "property": {
+            "required": False,
+            "choices": ["initial-network", "attach-post-create", "attach-post-deploy"],
```
Is there a reason to limit the set of properties to these three?

How about also supporting `static-web-listener` (not sure whether there are more)?properties

lemoony replied:
After skimming through code & documentation, i think these properties are also supported:

- bind-all-interfaces
- static-web-listener
- tld

I'm gonna add them.

ltalirz replied:
Thanks!

I leave it up to you whether you want to keep the list of properties here explicitly (which means we need to maintain the list in case dokku changes) or whether we simply allow any value and bubble up the error message to the user.

Whatever you decide, it may make sense to mention the list you now collected somewhere in the documentation of the module

lemoony replied:
Yes, I get your point. I think it makes to accept any string and let Dokku do the verification.

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,103 @@
+  dokku_network_property:
+    app: hello-world
+    network: example-network
+    property: attach-post-create
```
let's move the `property` above the value since, conceptually, you first need to know which property you are setting

ltalirz reviewed (changes requested):
Thanks a lot for the contribution @lemoony and sorry for the slight delay.

A couple of minor comments before we can get this merged.

ltalirz commented:
P.S. Would be great to add a short test to the [verify playbook](https://github.com/dokku/ansible-dokku/blob/master/molecule/default/verify.yml).

lemoony commented on library/dokku_network.py:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# from ansible.module_utils.basic import *
```
Can I ask why? All other modules also have the exact same lines. Also, isn't 

> #!/usr/bin/python

required for Ansible modules?

ltalirz replied:
Sure, I was just referring to the last line which is a commented import that is likely a leftover from previous edits.
My last PR did clean up the codebase a bit but if you encounter those, please feel free to delete them

lemoony replied:
Ah okay, I see. Makes sense.

lemoony commented:
Hi @ltalirz , thanks for the feedback! I already applied most of your suggestions. I will also try to add a test the next couple of days an then request a new review :)

ltalirz commented:
> Hi @ltalirz , thanks for the feedback! I already applied most of your suggestions. I will also try to add a test the next couple of days an then request a new review :)

Thanks a lot, much appreciated!

ltalirz commented on molecule/default/verify.yml:
```diff
@@ -73,3 +73,52 @@
+  # Testing dokku_network_property
+  - name: Setting a network property for an app
+    dokku_network_property:
+      app: hello-world
```
I think you want to go with `example-app` here, which is created in the lines above

lemoony replied:
Yeah, wasn't the only problem... I had a hard time getting molecule test getting to run on my local system properly. Neverless, tests are working now.

ltalirz replied:
> I had a hard time getting molecule test getting to run on my local system properly. Neverless, tests are working now.

Ah - I mostly just rely on the CI test setup myself.
What are the issues you ran into? Would be happy to add them to the [contributing docs](https://github.com/dokku/ansible-dokku/blob/master/CONTRIBUTING.md) in a separate PR.

lemoony replied:
The thing is that molecule does not tell my on my local setup what's actually wrong:

I```
NFO     Running default > verify
INFO     Running Ansible Verifier

PLAY [Verify] ******************************************************************

TASK [Gathering Facts] *********************************************************
ok: [instance]

TASK [Run role without running any tasks] **************************************
ERROR! Could not find specified file in role: tasks/init.yml

PLAY RECAP *********************************************************************
instance                   : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

CRITICAL Ansible return code was 2, command was: ansible-playbook --inventory /Users/lemoony/.cache/molecule/ansible-dokku/default/inventory --skip-tags molecule-notest,notest /Users/lemoony/projects/opensource/ansible-dokku/molecule/default/verify.yml
WARNING  An error occurred during the test sequence action: 'verify'. Cleaning up.
INFO     Running default > cleanup
WARNING  Skipping, cleanup playbook not configured.
INFO     Running default > destroy
```

So I have to push the code each time to see what's wrong...

lemoony commented:
Hi @ltalirz, from my point of view i've addressed every issue. I've added some basic tests for network management. If we want to have more thorough tests, maybe it worth the effort to split `verify.yml` up into multiple files (otherwise it gets to confusing i guess).

If you find the time, would be great to have another review on it. Thanks!

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,106 @@
+    has_changed = False
+    meta = {"present": False}
+
+    command = "dokku network:set {0} {1} {2}".format(
```
I suggest you check whether both `app` and `global` attributes are set and return an error if they are.

My feeling is that constructing these command strings with if blocks step by step is easier to understand and maintain, but feel free to keep this more compact notation.

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,106 @@
+    fields = {
+        "global": {"required": False, "default": False, "type": "bool"},
+        "app": {"required": False, "type": "str"},
+        "property": {"required": False, "type": "str"},
```
Property is required, isn't it?

```suggestion
        "property": {"required": True, "type": "str"},
```

ltalirz commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,106 @@
+        "global": {"required": False, "default": False, "type": "bool"},
+        "app": {"required": False, "type": "str"},
+        "property": {"required": False, "type": "str"},
+        "value": {"required": True, "type": "str"},
```
Value isn't required

```suggestion
        "value": {"required": False, "type": "str"},
```

Please just add one more task to the verify playbook that unsets a property (which would have caught this)

ltalirz reviewed (changes requested):
thanks a lot @lemoony 

almost ready to merge

lemoony commented on library/dokku_network_property.py:
```diff
@@ -0,0 +1,106 @@
+    command = "dokku network:set {0} {1} {2}".format(
+        "--global" if data["global"] else data["app"],
+        data["property"],
+        data["value"] if "value" in data else "",
```
`data["value"]` will only be executed by the engine if the condition `"value" in data` evalutes to true, otherwise the empty string value will be taken. I just tested this with a small python script and it should actually be fine.

lemoony replied:
Ahh sorry, I just realized what you may have pointed at: The above solution works if `value` is not present at all. However, if `value`is set to `None` the expression actually evaluates to true. 

I have to fix that. Thanks for spotting that!

ltalirz reviewed (approved):
thanks, all looking good now.

I see the tests pass on your fork - will merge

ltalirz commented:
And dropped new release https://github.com/dokku/ansible-dokku/releases/tag/v2021.9.3
